File: bindings/python/testvector.py
#!/usr/bin/env python

import libmapper as mpr
import signal

done = False

print('starting testvector.py')
print('libmapper version:', mpr.__version__, 'with' if mpr.has_numpy() else 'without', 'numpy support')

def handler_done(signum, frame):
    global done
    done = True

signal.signal(signal.SIGINT, handler_done)
signal.signal(signal.SIGTERM, handler_done)

def h(sig, event, id, val, time):
    print('  handler got', sig['name'], '=', val, 'at time', time.get_double())

def h2(sig, event, id, val, time):
    print('  handler2 got', sig['name'], '=', val, 'at time', time.get_double())

src = mpr.Device("py.testvector.src")
outsig = src.add_signal(mpr.Signal.Direction.OUTGOING, "outsig", 10, mpr.Type.INT32, None, 0, 1)

dest = mpr.Device("py.testvector.dst")
insig = dest.add_signal(mpr.Signal.Direction.INCOMING, "insig", 10, mpr.Type.FLOAT, None, 0, 1, None, h)

insig.set_callback(h2)

while not done and (not src.ready or not dest.ready):
    src.poll(10)
    dest.poll(10)

map = mpr.Map(outsig, insig)
map.push()

while not done and not map.ready:
    src.poll(10)
    dest.poll(10)

for i in range(100):
    if done:
        break
    outsig.set_value([i, i+1, i+2, i+3, i+4, i+5, i+6, i+7, i+8, i+9])
    dest.poll(10)
    src.poll(0)

print('freeing devices')
src.free()
dest.free()
print('done')

File: bindings/csharp/testvector.cs
using System;
using Mapper;

public class TestVector
{
    private static void SignalHandler(Signal sig, Mapper.Signal.Event evt, float[] value, Time time)
    {
        Console.WriteLine("Signal received value [" + String.Join(",", value) + "]");
    }

    public static void Main(string[] args)
    {
        Device dev = new Device("csharp.testvector");

        int[] min = {1,2,3,4}, max = {10,11,12,13};
        Signal outsig = dev.AddSignal(Signal.Direction.Outgoing, "outsig", 4, Mapper.Type.Float)
                           .SetProperty(Property.Min, min)
                           .SetProperty(Property.Max, max);
        Console.WriteLine("created signal outsig");

        Signal insig = dev.AddSignal(Signal.Direction.Incoming, "insig", 4, Mapper.Type.Float)
                          .SetCallback((Action<Signal, Signal.Event, float[], Time>)SignalHandler,
                                       Mapper.Signal.Event.Update);
        Console.WriteLine("created Signal insig");

        Console.Write("Waiting for Device...");
        while (dev.GetIsReady() == 0)
        {
            dev.Poll(25);
        }
        Console.WriteLine("ready: " + dev);

        Map map = new Map(outsig, insig);
        map.Push();

        Console.Write("Waiting for Map...");
        while (map.GetIsReady() == 0)
        {
            dev.Poll(25);
        }
        Console.WriteLine("ready!");

        float[] sig_val = {0.0F, 1.0F, 2.0F, 3.0F};
        int counter = 0;
        while (++counter < 100)
        {
            outsig.SetValue(sig_val);
            dev.Poll(100);
            for (int i = 0; i < 4; i++)
                sig_val[i] *= 1.1F;
            Console.WriteLine("Signal updated to [" + String.Join(",", sig_val) + "]");
        }
    }
}
